spec/Section 3 -- Type System.md

```diff
@@ -2164,3 +2164,128 @@ to the relevant IETF specification.
 ```graphql example
 scalar UUID @specifiedBy(url: "https://tools.ietf.org/html/rfc4122")
 ```
+
+## Schema Coordinates
+
+SchemaCoordinate :
+
+- Name
+- Name . Name
+- Name . Name ( Name : )
+- @ Name
+- @ Name ( Name : )
+
+:: A _schema coordinate_ is a human readable string that uniquely identifies a
+_schema element_ within a GraphQL Schema.
+
+:: A _schema element_ is a specific instance of a named type, field, input
+field, enum value, field argument, directive, or directive argument.
+
+A _schema coordinate_ is always unique. Each _schema element_ may be referenced
+by exactly one possible schema coordinate.
+
+A _schema coordinate_ may refer to either a defined or built-in _schema
+element_. For example, `String` and `@deprecated(reason:)` are both valid schema
+coordinates which refer to built-in schema elements. However it must not refer
+to a meta-field. For example, `Business.__typename` is _not_ a valid schema
+coordinate.
+
+Note: A union member is not a valid _schema coordinate_ as they reference
+existing types in the schema. This preserves the uniqueness property of a
+_schema coordinate_ as stated above.
+
+Note: A {SchemaCoordinate} is not a definition within a GraphQL {Document}, but
+a separate standalone grammar, intended to be used by tools to reference types,
+fields, and other *schema element*s. Examples include: as references within
+documentation to refer to types and fields in a schema, a lookup key that can be
+used in logging tools to track how often particular fields are queried in
+production.
+
+**Semantics**
+
+To refer to a _schema element_, a _schema coordinate_ must be interpreted in the
+context of a GraphQL {schema}.
+
+SchemaCoordinate : Name
+
+1. Let {typeName} be the value of the first {Name}.
+2. Return the type in the {schema} named {typeName}.
+
```

Comment:
Nit: should we assert that the type exists?

```suggestion
1. Let {typeName} be the value of the first {Name}.
2. Let {type} be the type in the {schema} named {typeName}.
3. Assert {type} exists.
4. Return {type}.

```

This is mainly for symmetry with the "argument coordinate" case below which asserts field existence and are therefore allowed to "fail".  Asserting that the type exists makes it explicit that a coordinate may "fail".

Reply:
I don't think we should do this. For example, if you wanted to ask the question "Does `Foo.bar(baz:)` exist in the schema, you might first check the schema coord `Foo` and make sure it's an object or interface, and then check `Foo.bar` exists, and finally access `Foo.bar(baz:)`. It should not be required that checking for `Foo` causes it to throw IMO - returning `null` seems not unreasonable. Throwing if you try to access `Foo.bar(baz:)` when `Foo.bar` doesn't exist does however seem reasonable.

Reply:
Does it even make sense to have "throw" or return "null"? I mean this is all declarative, it should be possible to represent an invalid coordinate if people want (for counter examples or whatever...). 

If we were to actually implement a function I would probably use a verb:

```
GetAstNode(SchemaCoordinate):
...
```

Do we want graphql-js to have such functions?

Reply:
That's effectively what we're doing

> To refer to a _schema element_, a _schema coordinate_ must be interpreted in the context of a GraphQL {schema}.

I.e. a schema coordinate can exist, but if you want to refer to a schema element then you follow the algorithm.

Reply:
Alright so what we're saying is:

```
throws if any component leading to the last component.
returns null if the last component of the coordinate is not found.
else return thee element.
```

If we're going that road, I would say we probably need to be explicit about it:

```
1. Let {typeName} be the value of the first {Name}.
2. If there is not type in the schema named {typeName} return null.
2. Else let {type} be the type in the {schema} named {typeName}.
4. return {type}.
```

Reply:
> it should be possible to represent an invalid coordinate if people want (for counter examples or whatever...)

At this risk of going down a rabbit hole of over-explicitness -- the spec as written states "a _schema coordinate_ must be interpreted in the context of a GraphQL {schema}." 

We are encoding parsing and validating within the same set of instructions here.

If we wanted to explicitly allow for "representing an invalid coordinate" without someone tutting at you, we _could_ expose seperate parse() and validate() functions in the graphql-js implementation, which would mirror this spec edit

Reply:
@magicmark I like this 👍 . It's also more consistent with the rest of the spec.

Reply:
After reading the [graphql-js PR](https://github.com/graphql/graphql-js/pull/3044/files), I think it'd make sense to have `resolveTypeCoordinate(TypeCoordinate)`, `resolveMemberCoordinate(MemberCoordinate)`, etc...

I can look into this for the next working group if there is interest, just let me know.